rfc/ipni-w3c.md

```diff
@@ -0,0 +1,346 @@
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    A[(dotstorage\nbucket)] -->|ObjectCreated fa:fa-car| B(bucket-to-indexer ƛ)
+    B -->|region/bucket/cid/cid.car| C[/indexer queue/]
+    C --> indexer(Indexer ƛ)
+    indexer --> |zQmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn| E[/multihash queue/]
+    E --> F(ipni Advertisement content ƛ)
+    F --> |PUT /advertCid|I
+    F --> |advert CID| G[/Advertisement queue/]
+    G --> H(ipni publish ƛ)
+    H --> |PUT /head|I[(Advert Bucket)]
+    H --> |POST head|IPNI[["`**IPNI**`"]]
+    
+    carpark[(carpark\nbucket)] --> |ObjectCreated fa:fa-car|w3infra-carpark-consumer(carpark-consumer ƛ)
+    w3infra-carpark-consumer -->|region/bucket/cid/cid.car| C[/indexer queue/]
+    indexer ---> dynamo[Dynamo\nblocks index]
+```
+
+## Proposal
+Provide a ipni/offer UCAN ability to submit the location of a CAR and a CARx-index that a user has stored with w3s, to make the them discoverable via IPFS implementations and other IPNI consumers.
+
+The ipni/offer is optional, and if not invoked the user's data is not indexed.
+
+### Indexing stored data
+
+```mermaid
+sequenceDiagram
+    actor Alice
+    Alice->>w3s: put CAR + CARx-index
+    activate Alice
+    Alice-->>Alice: create location claims
+    Alice->>w3s: ipni/offer (inc, loc claims)
+    deactivate Alice
+    activate w3s
+    w3s-->>w3s: fetch & verify CARx-index
+    w3s-->>w3s: create/store claims bucket
+    w3s-->>w3s: write advertisement
+    w3s-)ipni: publish head (CID)
+    deactivate w3s
+    ipni-->>w3s: fetch advert
+    activate ipni
+    ipni-->>ipni: index entries
+    deactivate ipni
+```
+Invoke `ipni/offer` with the CID for an [inclusion claim] that associates a CAR CID with a CARx-index CID.
+
+**UCAN invocation** example
+
+```json
+{
+  "iss": "did:key:zAlice",
+  "aud": "did:web:web3.storage",
+  "att": [{
+    "can": "ipni/offer",
+    "with": "did:key:space", // users space DID
+    "nb": {
+        "inclusion": CID   // inclusion claim CID
+    }
+  }]
+}
+```
+
+**Location claims** example
+
+```json
+{
+  "op": "assert/location",
+  "rsc": "https://web3.storage",
+  "input": {
+    "content" : CID /* CAR CID */, 
+    "location": "https://r2.cf/bag...car"
+  }
+}
+```
+```json
+{
+  "op": "assert/location",
+  "rsc": "https://web3.storage",
+  "input": {
+    "content" : CID /* CARx-index CID */, 
+    "location": "https://r2.cf/bag...carx"
+  }
+}
+```
+
+**Inclusion claim** example
+
+```json
+{
+  "content": CID, // CAR CID
+  "includes": CID // CARx-index Index CID
+}
+```
+
+When `ipni/offer` is invoked, encoded inclusion claim and encoded location claims for the CAR and the CARx-index are all sent with the invocation.
+
+_Alternatively, the w3s service could create the location claims. The user would still need to provide the locations, so might as well do this as a location claim. If w3s moves the data, then it can replace the location claims with new ones_
```

Comment:
right where it lives is owned by w3s, so w3s should create the locaiton claims -- but only on user trigger. I don't think the user needs to provide the locations per-se -- W3S is already tracking that.